raiden/smart_contracts/TokenNetwork.sol

```diff
@@ -0,0 +1,568 @@
+
+    event ChannelClosed(uint channel_identifier, address closing_address);
+
+    event ChannelUnlocked(uint channel_identifier, address participant, uint transferred_amount);
+
+    event TransferUpdated(uint channel_identifier, address caller);
+
+    event ChannelSettled(uint channel_identifier);
+
+    /*
+     * Modifiers
+     */
+
+    modifier isClosed(uint channel_identifier) {
+        require(closing_requests[channel_identifier].settle_block_number > 0);
+        _;
+    }
+
+    modifier stillTimeout(uint channel_identifier) {
+        require(closing_requests[channel_identifier].settle_block_number > block.number);
+        _;
+    }
+
+    modifier timeoutOver(uint channel_identifier) {
+        require(closing_requests[channel_identifier].settle_block_number <= block.number);
+        _;
+    }
+
+    // Note: we use the settleTimeout to check if the channel is open. It must always be > 0.
+    modifier settleTimeoutValid(uint timeout) {
+        require(timeout >= 6 && timeout <= 2700000);
+        _;
+    }
+
+    /*
+     *  Constructor
+     */
+
+    function TokenNetwork(address _token_address) public {
+        require(_token_address != 0x0);
+        require(contractExists(_token_address));
+
+        token = Token(_token_address);
+
+        // Check if the contract is indeed a token contract
+        require(token.totalSupply() > 0);
+    }
+
+    /*
+     *  Public functions
+     */
+
+    /// @notice Opens a new channel between `participant1` and `participant2`.
+    /// Can be called by anyone.
+    function openChannel(
+        address participant1,
+        address participant2,
+        uint settle_timeout)
+        settleTimeoutValid(settle_timeout)
+        public
+        returns (uint)
+    {
+        require(participant1 != 0x0);
+        require(participant2 != 0x0);
+        require(participant1 != participant2);
+
+        // Increase channel index counter
+        last_channel_index += 1;
+
+        require(channels[last_channel_index].settle_timeout == 0);
+        require(!channels[last_channel_index].participants[participant1].initialized);
+        require(!channels[last_channel_index].participants[participant2].initialized);
+
+        // Store channel information
+        channels[last_channel_index] = Channel({settle_timeout: settle_timeout});
+
+        // Mark the channel participants
+        // We use this in deposit to ensure the beneficiary is a channel participant
+        channels[last_channel_index].participants[participant1].initialized = true;
+        channels[last_channel_index].participants[participant2].initialized = true;
+
+        ChannelOpened(last_channel_index, participant1, participant2, settle_timeout);
+
+        return last_channel_index;
+    }
+
+    /// @notice Deposit tokens to an already existent channel.
+    /// Can be called by anyone.
+    function deposit(
+        uint channel_identifier,
+        address beneficiary,
+        uint256 added_amount)
```

Comment:
Not sure yet. I'm inclining to leave it like this and just add what you proposed: `openAndDeposit`

Reply:
please consider it, it make idempontent deposits possible.

Reply:
Having a hard time atm understanding what you want to say. I view `deposit` as having the `topUp` function. Even if called after the channel has been opened, it does `0 + added_amount`.

Reply:
I assumed topUp was idempontent, renaming is not important **idempotancy is.**

What I'm suggesting is the following:

```
function <name>(uint channel_identifier, address beneficiary, uint156 total_balance) {
        Channel storage channel = channels[channel_identifier];
        Participant storage participant = channel.participants[beneficiary];

        // Channel must be open and beneficiary must be one of the participants
        require(participant.initialized);

        // Channel cannot be closed
        require(closing_requests[channel_identifier].settle_block_number == 0);

        if (participant.balance < total_balance) {
            uint256 added_amount = participant.balance - total_balance;

            // Change the state
            channel.participants[beneficiary].balance += added_amount;

            // Do the transfer
            require(token.transferFrom(msg.sender, address(this), added_amount));

            ChannelNewBalance(channel_identifier, beneficiary, channel.participants[beneficiary].balance);
        }
}
```

Reply:
Ok, now I know what you mean. I understand the idea, but there are some issues that I see:

1. when "top up" is used, people think about `added_amount` - how much they are going to pay and not about `total_balance` - how much they will have in the end. When you go to top up / recharge your phone card, you don't tell&give the cashier  the `total_balance`, but the `added_amount`.

2. This amount of tokens from the constructor has to also be approved if using an ERC20 token. It's weird to have an approve transaction for `added_amount` and then another channel-related one with `total_balance`.

3. If you use something like ERC223, you have to provide the `added_amount`.

Bottom line, I think it's better to use `added_amount` and make sure the UI is nice and clear for the user and I think it makes more sense when you compare with everyday payments.

Reply:
1. Use a different name then.
2. it's doesnt make a different from the constructor perspective, since the initial balance is 0
3. the snippet above is using the `added_amount` for the token call

This is not about UI, it's about a property of the function that is useful. Consider this, you're writing an app that uses raiden's REST api, and the app wants to make a new deposit, but something happens, either the user closes and opens it again, the software crashes, the internet connectiong goes bad, and the REST response is never delivered.

Should the app make a new request? And what will be the excpected total balance if it doesnt, or after the second call?

Reply:
So, you want to replace the current `deposit` (added_amount) with a function that receives the `total_balance` as an argument.

(just to have an overview:) A has a balance of 3 tokens and wants to deposit 4 more.

1) ERC20:
- A calls `approve(TokenNetwork, 4)`
- A calls `deposit(channel_id, beneficiary, 7)`

2) ERC223:
- A calls `Token.transfer(TokenNetwork, 4, fallback_data)` -> token calls `TokenNetwork.fallback` -> `TokenNetwork.deposit(channel_id, beneficiary, 4 + beneficiary.balance)`

> the REST response is never delivered. Should the app make a new request?
On the client side you can watch for pending transactions and check A's balance after the transaction in mined. An issue would be network congestion and if you want to resend that transaction with a higher gas cost -> a bit more complicated here to send another transaction with the same nonce as the previous one.

So yes, handling this is easier with a `total_balance` approach because you can just send another separate transaction and not worry about ending up with a higher balance than A wanted.

Does it make sense to have both `topUp(added_amount)` and `setDeposit(total_amount)` in the contract?
Also @LefterisJP , please take a look over this.

Reply:
So there are 2 questions here.

One is the UI side, which I think we all agree should show the deposit amount you are about to deposit and not the total deposit we will have after deposit call succeeds.

On the API side having an idempotent call is a big plus. So I agree with Augusto that on our API side having a deposit of total amount function call makes sense. So something like `set_total_deposit()`.

> Does it make sense to have both topUp(added_amount) and setDeposit(total_amount)

I would say it does not make sense to have both.

Also the one problem I can see with this idempotency is that you will need to approve the contract to transfer tokens for you. So there you are going to have to use the actual amount of tokens about to top up and not the total.